### ISSUE TITLE
Poor AP and Detection results using faster R-CNN_inception_resnetV2_oid model

### ISSUE THREAD

**Issue by abhishek202206** [tensorflow/models#5125]
### System information
- **What is the top-level directory of the model you are using**:object detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Ubuntu 18.04
- **TensorFlow installed from (source or binary)**:Source
- **TensorFlow version (use command below)**:1.10
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:9.0/7.0
- **GPU model and memory**:Titax Xp
- **Exact command to reproduce**:

### Describe the problem
I have followed this article for using object detection on custom dataset https://blog.algorithmia.com/deep-dive-into-object-detection-with-open-images-using-tensorflow/. I have used google open images dataset and prepared the custom dataset as stated in the link. I used the model check point for inception_v2 model trained on OID. I trained the model for more than 142000 steps and the loss value also reduced substantially but the AP score for both classes have been extremely poor. 

![](https://screenshotscdn.firefoxusercontent.com/images/3158d6b2-7406-43ea-b4cc-35c1d186054b.png)

The config file is shown below.
model {
  faster_rcnn {
    num_classes: 2
    image_resizer {
      fixed_shape_resizer {
        height: 600
        width: 1024
      }
}
    feature_extractor {
      type: 'faster_rcnn_inception_resnet_v2'
      first_stage_features_stride: 16
    }
    first_stage_anchor_generator {
      grid_anchor_generator {
        scales: [0.25, 0.5, 1.0, 2.0]
        aspect_ratios: [0.5, 1.0, 2.0]
        height_stride: 8
        width_stride: 8
      }
    }
    first_stage_atrous_rate: 2
    first_stage_box_predictor_conv_hyperparams {
      op: CONV
      regularizer {
        l2_regularizer {
          weight: 0.0
        }
      }
      initializer {
        truncated_normal_initializer {
          stddev: 0.001
        }
      }
    }
    first_stage_nms_score_threshold: 0.0
    first_stage_nms_iou_threshold: 0.7
    first_stage_max_proposals: 300
    first_stage_localization_loss_weight: 2.0
    first_stage_objectness_loss_weight: 1.0
    initial_crop_size: 17
    maxpool_kernel_size: 1
    maxpool_stride: 1
    second_stage_box_predictor {
      mask_rcnn_box_predictor {
        fc_hyperparams {
          op: FC
          regularizer {
            l2_regularizer {
              weight: 0.0
            }
          }
          initializer {
            variance_scaling_initializer {
              factor: 1.0
              uniform: true
              mode: FAN_AVG
            }
          }
        }
        use_dropout: false
        dropout_keep_probability: 1.0
      }
    }
    second_stage_post_processing {
      batch_non_max_suppression {
        score_threshold: 0.300000011921
        iou_threshold: 0.600000023842
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SOFTMAX
    }
    second_stage_localization_loss_weight: 2.0
    second_stage_classification_loss_weight: 1.0
  }
}

train_config: {
  batch_size: 1
  optimizer {
   adam_optimizer: {
        learning_rate {
            exponential_decay_learning_rate: {initial_learning_rate:0.00001}
            }
        }
  }
  gradient_clipping_by_norm: 10.0
  fine_tune_checkpoint: "object_detection/faster_rcnn_resnet101_coco/model.ckpt"
  from_detection_checkpoint: true
  # Note: The below line limits the training process to 800K steps, which we
  # empirically found to be sufficient enough to train the Open Images dataset.
  # This effectively bypasses the learning rate schedule (the learning rate will
  # never decay). Remove the below line to train indefinitely.
  #num_steps: 8000000
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
}

train_input_reader: {
  tf_record_input_reader {
   input_path: "object_detection/data/train_window_door.record"
  }
  label_map_path: "object_detection/data/label_map_train_door_windows.pbtxt"
}

eval_config: {
  #metrics_set: "open_images_metrics"
  num_examples: 544
  metrics_set: "open_images_V2_detection_metrics"
  #batch_queue_capacity: 50
  # Note: The below line limits the evaluation process to 10 evaluations.
  # Remove the below line to evaluate indefinitely.
  max_evals: 10
}

eval_input_reader: {
  tf_record_input_reader {
    input_path: "object_detection/data/validate_window_door.record"
  }
  label_map_path: "object_detection/data/label_map_train_door_windows.pbtxt"
  shuffle: true
  num_readers: 1
}

Please have a look at the output snapshot. Every time I run the eval.py I get the same image no matter how many epoch the model runs. If you have a look the model shows poor detection even after so many epochs.  
![](https://screenshotscdn.firefoxusercontent.com/images/e47303fa-afe2-4fa7-a2dd-cdb544dcb82b.png)

Any help would be much appreciated.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by mawanda-jun** [comment]
Hi,
I think that your learning rate is too low. In addition, with the power that your hardware gives you I'd suggest you to set `batch_size` to 16 or 32, maybe 64 if your training set is made up of b/w jpeg images. This gives better result most of the times, even if it's a bit slow due to throughput.
For the configuration of your `learning_rate`, this is how I did: 
```
train_config: {
  batch_size: 4
  optimizer {
	# momentum_optimizer {
    adam_optimizer: {
      learning_rate: {
        exponential_decay_learning_rate {
          initial_learning_rate: 0.0001  # not so high: after 2 decay it's already at ~10e-5
          decay_steps: 600  # every 600 steps there is a decay
          decay_factor: 0.95  # I you look at the formula this gives you a good and not-so-fast descendant
        }
      }
      # momentum_optimizer_value: 0.9
    }
    use_moving_average: false
  }
  ...
```

Another point: why `score_treshold` and `iou_threshold` have those values? It's curiosity, since the default are not set like this.
Plus, I'd remove `max_evals` config to run evaluation indefinitely. You should put `eval_interval_secs: 150` so you have a new evaluation (aka new blue point on tensorboard) every 2.5min.
I hope those

**Comment by abhishek202206** [comment]
pe th



________________________________
